### ISSUE TITLE
AdaptiveTopK returns fully dense codes — _AdaptiveTopKSTE.forward STE base is `z`, not `z_soft`

### ISSUE THREAD

**Issue by SauersML** [SauersML/gam#353]
## Summary
`gamfit.torch.AdaptiveTopK` never sparsifies: its forward output keeps **all F** atoms nonzero regardless of `k_pred`. The learned per-row K (`k_pred`) is correct, but the masked activations are dense — so any SAE built on it has zero sparsity (silent; no error).

## Repro (gamfit 0.1.134, PyPI wheel)
```python
import torch
from gamfit.torch import AdaptiveTopK
torch.manual_seed(0)
g = AdaptiveTopK(F=64, k_min=4, k_max=16, head="linear", temperature=0.1)
z = torch.randn(8, 64)
z_active, k_pred = g(z)
print("k_pred per row:", [round(float(v), 1) for v in k_pred])  # ~11-12, OK (within [4,16])
print("nonzeros per row:", (z_active != 0).sum(-1).tolist())    # [64, 64, ...] !!
```
Output:
```
k_pred per row: [11.7, 11.8, 11.3, 11.9, 10.7, 11.5, 12.5, 10.5]
nonzeros per row: [64, 64, 64, 64, 64, 64, 64, 64]
```

## Expected
~`round(k_pred_i)` nonzeros per row — a hard per-row top-K forward, exactly as `_AdaptiveTopKSTE`'s own docstring promises ("hard per-row top-round(k_pred) mask forward, soft-mask gradient backward").

## Root cause
`gamfit/torch/modules.py`, `_AdaptiveTopKSTE.forward`:
```python
z_active_soft = z * soft_mask.to(z.dtype)
z_active = z + (z * hard_mask - z_active_soft).detach()   # <-- STE base is `z`
```
The straight-through value is `z + (z*hard - z*soft) = z*(1 + hard - soft)`, which is approximately `z` for every entry (top-K: hard=1, soft~=1 -> ~z; non-top-K: hard=0, soft~=0 -> ~z). Nothing is zeroed.

The STE should be based on the soft output so the forward value equals the hard-masked tensor:
```python
z_active = z_active_soft + (z * hard_mask - z_active_soft).detach()   # forward == z*hard_mask
```
Backward still flows through `z_active_soft` (the intended soft-mask gradient); only the forward value is corrected.

## Impact
Models using AdaptiveTopK for sparsity silently get dense codes. Found while building an SAE on the primitive; worked around downstream by re-imposing the hard mask, but the primitive is broken as shipped in 0.1.134.

**Comment by SauersML** [comment]
Closing — landed fix `7537de5ad` is correct and now has a real regression test.

**Root cause, principled.** `_AdaptiveTopKSTE.forward` (gamfit/torch/modules.py:125) changed the STE base from `z` to `z_active_soft`:
```python
z_active = z_active_soft + (z * hard_mask - z_active_soft).detach()
```
The old form `z + (z*hard - z_soft).detach()` evaluates numerically to `z*(1+hard-soft) ≈ z` for every atom (hard≈soft on both top-K and non-top-K entries), so nothing was zeroed and the output was dense. The fix makes the forward value exactly `z*hard_mask` (hard per-row top-round(k_pred)) while `.detach()` keeps `z_active_soft` as the differentiable carrier for the analytic soft-mask backward — matching the docstring contract. One-line, no shim.

**Regression test added** (`d1b703795`): `tests/test_adaptive_topk_hard_forward_issue_353.py` runs the exact issue repro (F=64, k_min=4, k_max=16, linear head) and asserts (1) nonzeros-per-row == round(k_pred_i), not 64; (2) the surviving entries equal `z*hard_mask` exactly (no soft attenuation of the kept atoms); and (3) backward still flows finite gradients to `z` and the k-head, confirming the soft carrier remains in the graph. Skips cleanly when torch is absent via `importorskip`.
